### PR TITLE
 #507309 added error handling pages

### PR DESCRIPTION
**Pull request by matkovskyi**
Added graphql service and added changes in initializer nextjs because there uses [[...path]].tsx in the nextjs-sxa module(need to reuse the prerender variable in the nextjs-sxa)

### CODE REVIEW

File: packages/sitecore-jss/src/debug.ts [hunk 1]
```diff
@@ -36,4 +36,5 @@ export default Object.freeze({
   robots: debug(`${rootNamespace}:robots`),
   redirects: debug(`${rootNamespace}:redirects`),
   personalize: debug(`${rootNamespace}:personalize`),
+  errohandling: debug(`${rootNamespace}:errohandling`),
```
Comment (illiakovalenko) [inline packages/sitecore-jss/src/debug.ts:39]: Typo `errohandling`

File: packages/sitecore-jss/src/site/graphql-error-handling-service.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    const errorHandlingResult: Promise<ErrorHandlingQueryResult> = this.graphQLClient.request(
+      this.query,
+      {
+        siteName,
+        language,
+      }
+    );
+    try {
```
Comment (illiakovalenko) [inline packages/sitecore-jss/src/site/graphql-error-handling-service.ts:83]: You could just write:
```js
return this.graphQLClient.request(
      this.query,
      {
        siteName,
        language,
      }
    ).then(...)
    .catch(...)
```

Try/catch is not needed here

File: packages/sitecore-jss/src/site/graphql-error-handling-service.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  /**
+   * Gets a GraphQL client that can make requests to the API. Uses graphql-request as the default
+   * library for fetching graphql data (@see GraphQLRequestClient). Override this method if you
+   * want to use something else.
+   * @returns {GraphQLClient} implementation
+   */
+  protected getGraphQLClient(): GraphQLClient {
+    return new GraphQLRequestClient(this.options.endpoint, {
+      apiKey: this.options.apiKey,
+      debugger: debug.errohandling,
```
Comment (illiakovalenko) [inline packages/sitecore-jss/src/site/graphql-error-handling-service.ts:101]: Typo `errohandling`

File: packages/create-sitecore-jss/src/templates/nextjs/src/lib/page-props.ts [hunk 4]
```diff
@@ -13,4 +14,5 @@ export type SitecorePageProps = {
   componentProps: ComponentPropsCollection;
   notFound: boolean;
   layoutData: LayoutServiceData;
+  errorHandlingPages?: ErrorHandlingType;
```
Comment (illiakovalenko) [inline packages/create-sitecore-jss/src/templates/nextjs/src/lib/page-props.ts:17]: If it's not used in the base app, it shouldn't be here, it's SXA related property

File: packages/create-sitecore-jss/src/templates/nextjs-sxa/src/pages/[[...path]].tsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+// It may be called again, on a serverless function, if
+// revalidation (or fallback) is enabled and a new request comes in.
+export const getStaticProps: GetStaticProps = async (context) => {
+<% } else if (prerender === 'SSR') { -%>
+// This function gets called at request time on server-side.
+export const getServerSideProps: GetServerSideProps = async (context) => {
+<% } -%>
+  const props = await sitecorePagePropsFactory.create(context);
+
+<% if (templates.includes('nextjs-sxa')) { -%>
```
Comment (illiakovalenko) [inline packages/create-sitecore-jss/src/templates/nextjs-sxa/src/pages/[[...path]].tsx:85]: You are inside the `nextjs-sxa` template, so it means that you should leave changes here which you need, don't need to check for sxa template :)

File: packages/create-sitecore-jss/src/init-runner.ts [hunk 6]
```diff
@@ -18,6 +18,10 @@ export const initRunner = async (initializers: string[], args: BaseArgs) => {
       args.silent || console.log(chalk.cyan(`Initializing '${init}'...`));
       const response = await initializer.init(args);
```
Comment (illiakovalenko) [inline packages/create-sitecore-jss/src/init-runner.ts:19]: I would suggest to use such approach:
```js
const { appName, nextSteps, initializers, ...rest } = await initializer.init(args);

args = { ...args, ...rest }
```

File: packages/create-sitecore-jss/src/common/args/base.ts [hunk 7]
```diff
@@ -29,6 +30,11 @@ export type BaseArgs = {
    * Default values will be used
    */
   yes?: boolean;
+  /**
+   * Use to prevent any questions related to the CLI argument values.
+   * Default values will be used
+   */
+  prerender?: Prerender;
```
Comment (illiakovalenko) [inline packages/create-sitecore-jss/src/common/args/base.ts:37]: This argument is only related to the Nextjs sample, I would suggest to add it to the existing `initializers/nextjs/args` file

File: packages/create-sitecore-jss/src/common/processes/transform.test.ts [hunk 8]
```diff
@@ -21,7 +22,7 @@ describe('transform', () => {
         appName: 'test',
         destination: '.\\test-data\\test',
         fetchWith: 'GraphQL',
-        prerender: 'SSG',
+        prerender: 'SSG' as Prerender,
```
Comment (illiakovalenko) [inline packages/create-sitecore-jss/src/common/processes/transform.test.ts:25]: Do we need to cast the type?

File: packages/create-sitecore-jss/src/common/Initializer.ts [hunk 9]
```diff
@@ -1,9 +1,10 @@
 import { BaseArgs } from './args/base';
-
+import { Prerender } from '../initializers/nextjs/prompts';
 export type InitializerResults = {
   appName: string;
   initializers?: string[];
   nextSteps?: string[];
+  prerender?: Prerender;
```
Comment (illiakovalenko) [inline packages/create-sitecore-jss/src/common/Initializer.ts:7]: The same issue here, `prerender` can't be returned by the each framework, only related to the Nextjs.
Probably, better to add `{ [key: string]: unknown }`

File: packages/create-sitecore-jss/src/common/args/base.ts (hunk 7 above)
Comment (illiakovalenko) [reply packages/create-sitecore-jss/src/common/args/base.ts:37]: Description is a copy from the prev prop, it's not correct

File: packages/sitecore-jss/src/site/graphql-error-handling-service.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+ * The schema of data returned in response to error pages link request
+ */
+export type ErrorHandlingQueryResult = { site: { siteInfo: { errorHandling: ErrorHandlingType } } };
+
+/**
+ * Service that fetch the error handling data using Sitecore's GraphQL API.
+ */
+export class GraphQLErrorHandlingService {
```
Comment (illiakovalenko) [inline packages/sitecore-jss/src/site/graphql-error-handling-service.ts:48]: I'm struggling to decide about the name, but for me seems like `ErrorHandling` is a very abstract name, we should have in this case something like:
`GraphQLErrorHandlingPagesService` and -> `graphql-error-handling-pages-service`

**Review by illiakovalenko — CHANGES_REQUESTED** [review]

File: packages/sitecore-jss/src/debug.ts (hunk 1 above)
Comment (illiakovalenko) [reply packages/sitecore-jss/src/debug.ts:39]: Same question about the name, it's very abstract

**Review by ambrauer — CHANGES_REQUESTED** [review]
Overall, I think we can avoid a lot of the `nextjs-sxa` overrides (`[[...path]].tsx`, and all the initializer prompts/runner logic) if we enhance the base `nextjs` template to handle `redirect` in page props.
* `nextjs` > `SitecorePageProps`: add `redirect: Redirect;`
* `nextjs` > `[[...path.tsx]]`: handle `redirect` appropriately in SSG/SSR contexts
* `nextjs-sxa`: add page-props-factory plugin e.g. `error-pages` that runs at the end (order = 3?). Here, you will do the logic currently being done in `[[...path]].tsx` and set `props.redirect` as needed.

Another thought - instead of redirect approach, would it be better to work more with the built-in [custom error page features](https://nextjs.org/docs/advanced-features/custom-error-pag)? See comments re: potential problems with redirect approach.

I know that we also have the error page items available on the schema (`notFoundPage` and `serverErrorPage`) where we could grab layout directly, not sure if that helps us though.

File: packages/create-sitecore-jss/src/templates/nextjs-sxa/src/pages/[[...path]].tsx [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  if ((context.res.statusCode >= 500 || context.res.statusCode <= 511) && props.errorHandlingPages.serverErrorPagePath) {
+    return {
+      redirect: {
```
[the remark this answers is not in the document]
Comment (matkovskyi) [reply packages/create-sitecore-jss/src/templates/nextjs-sxa/src/pages/[[...path]].tsx:88]: Fixed! Yep, there need to use '&&' operator in the conditional

File: packages/sitecore-jss/src/site/graphql-error-handling-pages-service.ts [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        errorHandling(language: $language) {
+          notFoundPagePath
+          serverErrorPagePath
+        }
+      }
+    }
+  }
+`;
+
+export type GraphQLErrorHandlingPagesServiceConfig = {
```
Comment (ambrauer) [inline packages/sitecore-jss/src/site/graphql-error-handling-pages-service.ts:19]: Can we simplify the naming even more in these new services/types and get rid of "Handling"? i.e. `GraphQLErrorPagesServiceConfig` here

File: packages/sitecore-jss/src/site/graphql-error-handling-pages-service.ts [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+   * The JSS application name
+   */
+  siteName: string;
+  /**
+   * The language
+   */
+  language: string;
+};
+
+export type ErrorHandlingPagesType = { notFoundPagePath: string; serverErrorPagePath: string };
```
Comment (ambrauer) [inline packages/sitecore-jss/src/site/graphql-error-handling-pages-service.ts:38]: "Type" seems redundant
```suggestion
export type ErrorPages = { notFoundPagePath: string; serverErrorPagePath: string };
```

File: packages/create-sitecore-jss/src/templates/nextjs-sxa/src/pages/[[...path]].tsx [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+// It may be called again, on a serverless function, if
+// revalidation (or fallback) is enabled and a new request comes in.
+export const getStaticProps: GetStaticProps = async (context) => {
+<% } else if (prerender === 'SSR') { -%>
+// This function gets called at request time on server-side.
+export const getServerSideProps: GetServerSideProps = async (context) => {
+<% } -%>
+  const props = await sitecorePagePropsFactory.create(context);
+
+<% if (prerender === 'SSR') { -%>
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/templates/nextjs-sxa/src/pages/[[...path]].tsx:85]: I see you added the SSR check, but what's our story for SSG (build-time) custom error pages then?

**Comment by ambrauer** [comment]
@matkovskyi any thoughts on at least the first portion of the comment above?

> Overall, I think we can avoid a lot of the `nextjs-sxa` overrides (`[[...path]].tsx`, and all the initializer prompts/runner logic) if we enhance the base `nextjs` template to handle `redirect` in page props.
> 
> * `nextjs` > `SitecorePageProps`: add `redirect: Redirect;`
> * `nextjs` > `[[...path.tsx]]`: handle `redirect` appropriately in SSG/SSR contexts
> * `nextjs-sxa`: add page-props-factory plugin e.g. `error-pages` that runs at the end (order = 3?). Here, you will do the logic currently being done in `[[...path]].tsx` and set `props.redirect` as needed.

**Comment by matkovskyi** [comment]
> @matkovskyi any thoughts on at least the first portion of the comment above?
> 
> > Overall, I think we can avoid a lot of the `nextjs-sxa` overrides (`[[...path]].tsx`, and all the initializer prompts/runner logic) if we enhance the base `nextjs` template to handle `redirect` in page props.
> > 
> > * `nextjs` > `SitecorePageProps`: add `redirect: Redirect;`
> > * `nextjs` > `[[...path.tsx]]`: handle `redirect` appropriately in SSG/SSR contexts
> > * `nextjs-sxa`: add page-props-factory plugin e.g. `error-pages` that runs at the end (order = 3?). Here, you will do the logic currently being done in `[[...path]].tsx` and set `props.redirect` as needed.

I tried to use redirect into SitcorePageProps but I can see error
 _`Error serializing `.redirect` returned from `getServerSideProps` in "/[[...path]]".
Reason: `function` cannot be serialized as JSON. Please only return JSON serializable data types.`_

How I understand we cannot use some methods in Props data.

File: packages/create-sitecore-jss/src/templates/nextjs/src/pages/[[...path]].tsx [hunk 15]
```diff
@@ -91,6 +91,13 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
 <% } -%>
   const props = await sitecorePagePropsFactory.create(context);
 
+  // generic redirect - handler error pages
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/templates/nextjs/src/pages/[[...path]].tsx:94]:
```suggestion
  // Check if we have a redirect (e.g. custom error page)
```

File: packages/sitecore-jss/src/site/graphql-error-pages-service.ts [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+   * The JSS application name
+   */
+  siteName: string;
+  /**
+   * The language
+   */
+  language: string;
+};
+
+export type ErrorPagesType = { notFoundPagePath: string; serverErrorPagePath: string };
```
Comment (ambrauer) [inline packages/sitecore-jss/src/site/graphql-error-pages-service.ts:38]: Please add a JSDoc comment (since this is exposed publicly)

Comment (ambrauer) [inline packages/sitecore-jss/src/site/graphql-error-pages-service.ts:38]: Seems we can drop "Type" (redundant)?
```suggestion
export type ErrorPages = { notFoundPagePath: string; serverErrorPagePath: string };
```

File: packages/sitecore-jss/src/site/graphql-error-pages-service.ts [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+/**
+ * The schema of data returned in response to error pages link request
+ */
+type ErrorPagesQueryResult = {
+  site: { siteInfo: { errorHandling: ErrorPagesType } };
+};
+
+/**
+ * Service that fetch the error handling data using Sitecore's GraphQL API.
```
Comment (ambrauer) [inline packages/sitecore-jss/src/site/graphql-error-pages-service.ts:48]: Minor, but there are some uses of the old naming "handling" still being used in comments. Should be changed to "pages".

File: packages/create-sitecore-jss/src/init-runner.ts [hunk 18]
```diff
@@ -5,7 +5,7 @@ import { InitializerFactory } from './InitializerFactory';
 
 export const initRunner = async (initializers: string[], args: BaseArgs) => {
   let nextStepsArr: string[] = [];
-  let appName;
+  let nextAppName;
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/init-runner.ts:8]: Can this file be reverted? Doesn't look like there are any changes other than stylistically

File: packages/create-sitecore-jss/src/templates/nextjs-sxa/src/lib/page-props-factory/plugins/error-pages.ts [hunk 19]
```diff
@@ -0,0 +1,44 @@
+import { GetServerSidePropsContext, GetStaticPropsContext } from 'next';
+import { SitecorePageProps } from 'lib/page-props';
+import { Plugin } from 'lib/page-props-factory';
+import { GraphQLErrorPagesService } from '@sitecore-jss/sitecore-jss-nextjs';
+import { isServerSidePropsContext } from '..';
+import config from 'temp/config';
+
+class ErrorPagesPlugin implements Plugin {
+  order = 2;
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/templates/nextjs-sxa/src/lib/page-props-factory/plugins/error-pages.ts:9]: Let's bump this to 3 to be safe (personalize is 2 as well)
```suggestion
  order = 3;
```

File: packages/create-sitecore-jss/src/templates/nextjs/src/lib/page-props.ts [hunk 20]
```diff
@@ -13,4 +14,5 @@ export type SitecorePageProps = {
   componentProps: ComponentPropsCollection;
   notFound: boolean;
   layoutData: LayoutServiceData;
+  redirect: Redirect;
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/templates/nextjs/src/lib/page-props.ts:17]: Should this be optional?

File: packages/create-sitecore-jss/src/templates/nextjs-sxa/src/lib/page-props-factory/plugins/error-pages.ts [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+  async exec(props: SitecorePageProps, context: GetServerSidePropsContext | GetStaticPropsContext) {
+    const errorPagesService = new GraphQLErrorPagesService({
+      endpoint: config.graphQLEndpoint,
+      apiKey: config.sitecoreApiKey,
+      siteName: config.jssAppName,
+      language: props.locale,
+    });
+    const resultErrorPages = await errorPagesService.fetchErrorPages();
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/templates/nextjs-sxa/src/lib/page-props-factory/plugins/error-pages.ts:18]: We should only fetch these if there is a reason to (for performance reasons). Before this, we should first check for error state (i.e. notFound or 5xx), and exit if all good.

**Review by ambrauer — CHANGES_REQUESTED** [review]
See comments. This is looking great overall...

File: packages/create-sitecore-jss/src/templates/nextjs-sxa/src/lib/page-props-factory/plugins/error-pages.ts (hunk 21 above)
Comment (matkovskyi) [reply packages/create-sitecore-jss/src/templates/nextjs-sxa/src/lib/page-props-factory/plugins/error-pages.ts:18]: Yes, I agreed

**Review by ambrauer — APPROVED** [review]
Looks good!

**Review by illiakovalenko — APPROVED** [review]

**Pull request merged**